README.md
# SPID Auth Docker

SPID Auth Docker is a service provider based on shibboleth that takes care of
the SPID authentication process with the Identity Provider and redirects the
attributes of the final response to the set address so that they can be used
by the application.

SPID Auth Docker has been developed and is maintained by AgID - Agenzia per
l'Italia Digitale.

## Disclaimer

It's highly recommended to use the [latest release] of the image.
The versioning follows the [SemVer rules].

[latest release]: https://github.com/italia/spid-shibboleth-proxy-docker/tree/v1.1.1
[SemVer rules]: https://semver.org

## How to quickly test it

1.  Create a directory to store SAML certificates and another to store log
    files

        $ mkdir -vp /tmp/certs /tmp/log/{httpd,shibboleth,shibboleth-www}

2.  Create a docker compose file (`docker-compose.yml`) with the following
    content. **Note:** set `SERVER_NAME` to a publicly reachable IP or FQDN.

    ```.yaml
    version: '3'

    services:
      authproxy:
        image: spid-auth-proxy
        build:
          context: .
          dockerfile: Dockerfile
        ports:
          - '80:80'
          - '443:443'
        volumes:
          - '/tmp/certs:/opt/shibboleth-sp/certs'
          - '/tmp/log:/var/log'
        environment:
          ORGANIZATION: 'My eGov Service'
          SERVER_NAME: 'my.auth.proxy.com'
          ENTITY_ID: 'https://my.auth.proxy.com'
          TARGET_BACKEND: 'https://mytargetapp.my.cloud.provider.com'
          TARGET_LOCATION: '/login'
          ACS_INDEXES: '1;2'
          ACS_1_LABEL: 'set 1'
          ACS_1_ATTRS: 'name;familyName;fiscalNumber;email'
          ACS_2_LABEL: 'set 2'
          ACS_2_ATTRS: 'spidCode;fiscalNumber'
    ```

    You can use `docker-compose.quickstart.yml`, which is available in this
    repository.

3.  Execute the Docker environmnet with the following command

        $ docker-compose up --build

    It will

    *   build the docker image

    *   generate a self-signed certificate for TLS (`SERVER_NAME` used
        as `CN`), SAML assertions signature (`SAML Signature` as `CN`) and
        SAML metadata signature (`SAML Metadata Signature` as `CN`)

    *   generate the Shibboleth/Httpd configuration according to the
        environment variables

    *   execure `shibd` and `httpd`

4.  If everything gone well, you should be able to access the URL

        https://<SERVER_NAME>/metadata

    Open it and use the following information

    *   `entityID` under `<md:EntityDescriptor>` tag
    *   `<ds:X509Certificate>` under `<md:SPSSODescriptor>` tag
    *   `Location` under `<md:SingleLogoutService>` tag
    *   `Location` under `<md:AssertionConsumerService>` tag
    *   attributes names under `<md:AttributeConsumingService>` tags

    to register you authentication proxy on

        https://idp.spid.gov.it:8080

5.  Open the URL

        https://<SERVER_NAME>/access

    and click on

       Test on /whoami (lucia/password123)

    It starts an authentication process that will end on `/whoami` endpoint
    where all the information about the authentication will be dumped.

## How to use it in production-like environment

1.  Create a directory to store SAML certificates and another to store log
    files

        $ mkdir -vp \
            /opt/authproxy/certs/{saml,tls} \
            /opt/authproxy/log/{httpd,shibboleth,shibboleth-www}

2.  Create/Obtain X509 certificates for TLS and SAML signatures and store them
    as follows

        /opt/authproxy/certs
        ├── saml
        │   ├── sp-cert.pem
        │   ├── sp-key.pem
        │   ├── sp-meta-cert.pem
        │   └── sp-meta-key.pem
        └── tls
            ├── server.crt
            └── server.key

3.  Create a docker compose file as follows. Be sure to set environment
    variables to values reflecting your real environment.

    ```.yaml
    version: '3'

    services:
      authproxy:
        image: spid-auth-proxy
        build:
          context: .
          dockerfile: Dockerfile
        ports:
          - '80:80'
          - '443:443'
        volumes:
          - '/opt/authproxy/certs/saml:/opt/shibboleth-sp/certs:ro'
          - '/opt/authproxy/certs/tls/server.crt:/etc/pki/tls/certs/server.crt:ro'
          - '/opt/authproxy/certs/tls/server.key:/etc/pki/tls/private/server.key:ro'
          - '/opt/authproxy/log:/var/log'
        environment:
          MODE: 'prod'
          ORGANIZATION: 'My eGov Service'
          SERVER_NAME: 'my.auth.proxy.com'
          ERROR_URL: 'https://my.auth.proxy.com/error'
          ENTITY_ID: 'https://my.auth.proxy.com'
          TARGET_BACKEND: 'https://mytargetapp.my.cloud.provider.com'
          TARGET_LOCATION: '/login'
          ACS_INDEXES: '1;2'
          ACS_1_LABEL: 'set 1'
          ACS_1_ATTRS: 'name;familyName;fiscalNumber;email'
          ACS_2_LABEL: 'set 2'
          ACS_2_ATTRS: 'spidCode;fiscalNumber'
    ```

    Be sure that `MODE` envvar is set to `prod`.

    The URL specified in `ERROR_URL` should be an endpoint that is able
    to manage errors as mentioned in

        https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPErrors#NativeSPErrors-Redirection

4.  If necessary, revise the `httpd` configuration files under `etc/httpd/conf.d`
    in order to fit with your requirements

5.  Execute the Docker environmnet with the following command

        $ docker-compose up --build

6.  Register your authentication proxy on the IdP by providing the information
    contained at

        https://<SERVER_NAME>/metadata

7.  In order to use the authentication proxy, you application should
    initialise the the authentication by calling

        https://<SERVER_NAME>/iam/Login?target=https://<SERVER_NAME>/login&entityID=<IDP ENTITY_ID>

    Once authenticated, the callback (`/login`) will proxy the response to
    your backend (`TARGET_BACKEND`) by including within the request headers
    the authentication result.

## How to define AttributeConsumingService elements

The `AttributeConsumingService` (ACS) elements can be defined by using a set
of properly named environment variables.

Firstly, you have to declare the indexes of your ACSs by defining the
`ACS_INDEXES` variable with all the indexes separated by a `;`.

Then, for each ACS, you have to define the label with `ACS_<INDEX>_LABEL` and
the list of the attributes with `ACS_<INDEX>_ATTRS`. The attributes must be
specified as `;` separated list.

For instance, to the define two ACSs (with index `1` and `27`) where the
first includes the attributes `spidCode` and `fiscalNumber` while the latter
`name` and `placeOfBirth`, the environment variables must be defined and set
as follows

```.yaml
environment:
  ACS_INDEXES: '1;27'
  ACS_1_LABEL: 'My First Set'
  ACS_1_ATTRS: 'spidCode;fiscalNumber'
  ACS_27_LABEL: 'My Second Set'
  ACS_27_ATTRS: 'name;placeOfBirth'
```

The `ACS_*` environment variables are also used to configure the [`AttributeChecker`](https://wiki.shibboleth.net/confluence/display/SP3/Attribute+Checker+Handler) handler.
The environment variables of the example will generate the following configuration
(nested in the resulting `/etc/shibboleth/shibboleth2.xml`)

```.xml
<!-- Check the returned attributes -->
<Handler type="AttributeChecker" Location="/AttrChecker" template="attrChecker.html" flushSession="true">
    <AND>
        <OR>
            <Rule require="authnContextClassRef">https://www.spid.gov.it/SpidL1</Rule>
            <Rule require="authnContextClassRef">https://www.spid.gov.it/SpidL2</Rule>
            <Rule require="authnContextClassRef">https://www.spid.gov.it/SpidL3</Rule>
        </OR>
        <OR>
            <!-- Check AttributeConsumingService with index 1 -->
            <AND>
                <AND>
                    <Rule require="SPIDCODE"/>
                    <Rule require="FISCALNUMBER"/>
                </AND>
                <AND>
                    <NOT><Rule require="ADDRESS"/></NOT>
                    <NOT><Rule require="COMPANYNAME"/></NOT>
                    <NOT><Rule require="COUNTYOFBIRTH"/></NOT>
                    <NOT><Rule require="DATEOFBIRTH"/></NOT>
                    <NOT><Rule require="DIGITALADDRESS"/></NOT>
                    <NOT><Rule require="EMAIL"/></NOT>
                    <NOT><Rule require="EXPIRATIONDATE"/></NOT>
                    <NOT><Rule require="FAMILYNAME"/></NOT>
                    <NOT><Rule require="GENDER"/></NOT>
                    <NOT><Rule require="IDCARD"/></NOT>
                    <NOT><Rule require="IVACODE"/></NOT>
                    <NOT><Rule require="MOBILEPHONE"/></NOT>
                    <NOT><Rule require="NAME"/></NOT>
                    <NOT><Rule require="PLACEOFBIRTH"/></NOT>
                    <NOT><Rule require="REGISTEREDOFFICE"/></NOT>
                </AND>
            </AND>
            <!-- Check AttributeConsumingService with index 27 -->
            <AND>
                <AND>
                    <Rule require="NAME"/>
                    <Rule require="PLACEOFBIRTH"/>
                </AND>
                <AND>
                    <NOT><Rule require="ADDRESS"/></NOT>
                    <NOT><Rule require="COMPANYNAME"/></NOT>
                    <NOT><Rule require="COUNTYOFBIRTH"/></NOT>
                    <NOT><Rule require="DATEOFBIRTH"/></NOT>
                    <NOT><Rule require="DIGITALADDRESS"/></NOT>
                    <NOT><Rule require="EMAIL"/></NOT>
                    <NOT><Rule require="EXPIRATIONDATE"/></NOT>
                    <NOT><Rule require="FAMILYNAME"/></NOT>
                    <NOT><Rule require="FISCALNUMBER"/></NOT>
                    <NOT><Rule require="GENDER"/></NOT>
                    <NOT><Rule require="IDCARD"/></NOT>
                    <NOT><Rule require="IVACODE"/></NOT>
                    <NOT><Rule require="MOBILEPHONE"/></NOT>
                    <NOT><Rule require="REGISTEREDOFFICE"/></NOT>
                    <NOT><Rule require="SPIDCODE"/></NOT>
                </AND>
            </AND>
        </OR>
    </AND>
</Handler>
```

Furthermore, `ACS_*` variables are used to generate [`SessionInitiator`](#https://wiki.shibboleth.net/confluence/display/SP3/SessionInitiator) elements.
The environment variables of the example will generate the following configuration
(nested in the resulting `/etc/shibboleth/shibboleth2.xml`)

```.xml
<!-- SessionInitiator for AttributeConsumingService 0 -->
<SessionInitiator type="SAML2"
    Location="/Login0"
    isDefault="true"
    entityID="https://sp.example.com"
    outgoingBinding="urn:oasis:names:tc:SAML:profiles:SSO:request-init"
    isPassive="false"
    signing="true">
    <samlp:AuthnRequest xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
        xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
        Version="2.0" ID="placeholder0.example.com" IssueInstant="1970-01-01T00:00:00Z"
        AttributeConsumingServiceIndex="0" ForceAuthn="true">
        <saml:Issuer
            Format="urn:oasis:names:tc:SAML:2.0:nameid-format:entity"
            NameQualifier="https://sp.example.com">
            https://sp.example.com
        </saml:Issuer>
        <samlp:NameIDPolicy
            Format="urn:oasis:names:tc:SAML:2.0:nameid-format:transient"
        />
    </samlp:AuthnRequest>
</SessionInitiator>

<!-- SessionInitiator for AttributeConsumingService 27 -->
<SessionInitiator type="SAML2"
    Location="/Login27"
    isDefault="true"
    entityID="https://sp.example.com"
    outgoingBinding="urn:oasis:names:tc:SAML:profiles:SSO:request-init"
    isPassive="false"
    signing="true">
    <samlp:AuthnRequest xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
        xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
        Version="2.0" ID="placeholder27.example.com" IssueInstant="1970-01-01T00:00:00Z"
        AttributeConsumingServiceIndex="27" ForceAuthn="true">
        <saml:Issuer
            Format="urn:oasis:names:tc:SAML:2.0:nameid-format:entity"
            NameQualifier="https://sp.example.com">
            https://sp.example.com
        </saml:Issuer>
        <samlp:NameIDPolicy
            Format="urn:oasis:names:tc:SAML:2.0:nameid-format:transient"
        />
    </samlp:AuthnRequest>
</SessionInitiator>
```

With this mechanism, you can dynamically specify the `AttributeConsumingService` by using the URLs `/iam/Login0`, `/iam/Login27` and so on. For instance

```
https://sp.example.com/iam/Login0?target=https://sp.example.com/login&entityID=https://idp.spid.gov.it
                           ^^^^^^
```


CHANGELOG.md
# Changelog

## devel

## 1.1.1 (2019-09-18)

* Limit the redirect during the return path

## 1.1.0 (2019-01-13)

* Update SPID Registry signing certificate (fingerprint `CF:6F:E5:4E:9A:78:1A:F2:78:92:69:0B:A7:BF:FB:B1:8A:D1:28:B5:4D:40:06:DF:4B:06:8A:D1:4B:6D:EB:27`)
* Refine checks on returned attributes

## 1.0.0 (2018-12-18)

* Generate `<SessionInitiator>` elements according to `ACS_*` environment variables
* Add check on returned AuthnContextClassRef and attributes
* Add backend-side checks
* Update SPID registry certificate
* Make logging and security policy configurable
* Revise Shibboleth SP configuration (align to 3.x branch)
* Enable SPID SP validator
* Enable the use of `TARGET_LOCATION` environment variable
* Generate the `<AttributeConsumingService>` elements according to the
  defined environment variables
* Remove `AllowCreate` from `NameIDPolicy` according to
  [SPID - Avviso 5](https://www.agid.gov.it/sites/default/files/repository_files/documentazione/spid-avviso-n5-regole-tecniche-errata-corrige.pdf)
* Revise httpd security configuration (TLS, security headers)

## 0.4.0 (2018-07-06)

* Fix the order of SP metadata elements
* Substitute `samlsign` with `xmlsectool`
* Update signing (rsa-sha512) and digest (sha512) algorithm
* Remove double KeyName in SP requests signature
* Fix Shibboleth SP configuration
* Hack the metagen.sh script (cloned under /usr/local/bin) in order to match
  the SPID requirements

## 0.3.1 (2018-06-22)

* Fix metadata signing process

## 0.3.0 (2018-06-18)

* Add script for system integration
* Allow to run the container in production mode
* Add checking of the aggregate IdP metadata signature

## 0.2.0 (2018-05-24)

* Declare only SAML 2.0 as supported protocol
* Include organization details in metadata
* Fix typo in metadata tranformation (see #1)

## 0.1.1 (2018-05-23)

* Fix IdP metadata aggregator URI in Shibboleth SP configuration

## 0.1.0 (2018-05-17)

* First release


system-integration.sh
#!/bin/bash

# Copyright 2018 AgID - Agenzia per l'Italia Digitale
#
# Licensed under the EUPL, Version 1.2 or - as soon they will be approved by
# the European Commission - subsequent versions of the EUPL (the "Licence").
#
# You may not use this work except in compliance with the Licence.
#
# You may obtain a copy of the Licence at:
#
#    https://joinup.ec.europa.eu/software/page/eupl
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the Licence is distributed on an "AS IS" basis, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# Licence for the specific language governing permissions and limitations
# under the Licence.

_INSTALL_DIR=${INSTALL_DIR:-"/opt/authproxy"}

if [ ! -d "${_INSTALL_DIR}/spid-auth-docker" ]; then
    cat<<EOF
ERROR: INSTALL_DIR must be set to the path where
       https://github.com/italia/spid-auth-docker has been
       cloned.

       For instance, ff you cloned the repository under
       "/tmp/spid-auth-docker", INSTALL_DIR must be set
       to "/tmp".
EOF
    exit 1
fi

_SERVER_NAME=${SERVER_NAME:-"my.auth.proxy.it"}
_ENTITY_ID=${ENTITY_ID:-"https://my.auth.proxy.it"}
_TARGET_BACKEND=${TARGET_BACKEND:-"http://backend_app:8080"}
_TARGET_LOCATION=${TARGET_LOCATION:-"/login"}
_ORGANIZATION=${ORGANIZATION:-"A Company Making Everything"}
_MODE=${MODE:-"dev"}

LOG_DIR="${_INSTALL_DIR}/log"
CERT_DIR="${_INSTALL_DIR}/certs"
LOGROTATE_DIR="${_INSTALL_DIR}/etc/logrotate.d"
SYSTEMD_DIR="${_INSTALL_DIR}/lib/systemd/system"

# create log dirs
mkdir -vp ${LOG_DIR}/{httpd,shibboleth,shibboleth-www}

# create cert dirs
mkdir -vp ${CERT_DIR}/{saml,tls}

# create system dirs
mkdir -vp ${LOGROTATE_DIR} ${SYSTEMD_DIR}

#
# generate docker compose file
#
cat > ${_INSTALL_DIR}/docker-compose.yml <<EOF
version: '3'
services:
  authproxy:
    image: spid-auth-proxy
    restart: always
    build:
      context: ${_INSTALL_DIR}/spid-auth-docker
      dockerfile: Dockerfile
      args:
        http_proxy: ${http_proxy}
        https_proxy: ${https_proxy}
    ports:
      - '80:80'
      - '443:443'
    volumes:
      - '${CERT_DIR}/saml:/opt/shibboleth-sp/certs'
      - '${CERT_DIR}/tls/server.crt:/etc/pki/tls/certs/server.crt:ro'
      - '${CERT_DIR}/tls/server.key:/etc/pki/tls/private/server.key:ro'
      - '${LOG_DIR}:/var/log'
    environment:
      MODE: '${_MODE}'
      ORGANIZATION: '${_ORGANIZATION}'
      SERVER_NAME: '${_SERVER_NAME}'
      ENTITY_ID: '${_ENTITY_ID}'
      TARGET_BACKEND: '${_TARGET_BACKEND}'
      TARGET_LOCATION: '${_TARGET_LOCATION}'
      SPID_ACS: |
        <md:AttributeConsumingService index="0">
          <md:ServiceName xml:lang="it">all SPID attributes</md:ServiceName>
          <md:RequestedAttribute Name="address" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"/>
          <md:RequestedAttribute Name="companyName" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"/>
          <md:RequestedAttribute Name="countyOfBirth" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"/>
          <md:RequestedAttribute Name="dateOfBirth" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"/>
          <md:RequestedAttribute Name="digitalAddress" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"/>
          <md:RequestedAttribute Name="email" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"/>
          <md:RequestedAttribute Name="expirationDate" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"/>
          <md:RequestedAttribute Name="familyName" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"/>
          <md:RequestedAttribute Name="fiscalNumber" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"/>
          <md:RequestedAttribute Name="gender" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"/>
          <md:RequestedAttribute Name="idCard" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"/>
          <md:RequestedAttribute Name="ivaCode" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"/>
          <md:RequestedAttribute Name="mobilePhone" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"/>
          <md:RequestedAttribute Name="name" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"/>
          <md:RequestedAttribute Name="placeOfBirth" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"/>
          <md:RequestedAttribute Name="registeredOffice" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"/>
          <md:RequestedAttribute Name="spidCode" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"/>
        </md:AttributeConsumingService>
    networks:
      - authproxy
    depends_on:
      - backend_app
  backend_app:
    build:
      context: ${_INSTALL_DIR}/spid-auth-docker/example
      dockerfile: Dockerfile
      args:
        http_proxy: ${http_proxy}
        https_proxy: ${https_proxy}
    environment:
      SERVER_NAME: ${_SERVER_NAME}
    networks:
      - authproxy
networks:
  authproxy:
    driver_opts:
      com.docker.network.driver.mtu: 1500
EOF

#
# generate systemd unit file
#
cat > ${SYSTEMD_DIR}/spid-auth-proxy.service <<EOF
[Unit]
Description=SPID Authentication Proxy
Requires=docker.service
After=docker.service
[Service]
Restart=always
WorkingDirectory=${_INSTALL_DIR}
# stop and remove existing containers
ExecStartPre=/usr/local/bin/docker-compose stop
ExecStartPre=/usr/local/bin/docker-compose rm -f
# start
ExecStart=/usr/local/bin/docker-compose up --build
# stop
ExecStop=/usr/local/bin/docker-compose stop
[Install]
WantedBy=multi-user.target
EOF

cat <<EOF

Run the following command to enable the systemd unit file
    cd /lib/systemd/system
    sudo ln -s ${SYSTEMD_DIR}/spid-auth-proxy.service .
    sudo systemctl enable spid-auth-proxy.service

EOF

#
# generate logrotate file
#
cat > ${LOGROTATE_DIR}/spid-auth-proxy <<EOF
${LOG_DIR}/httpd/*_log
${LOG_DIR}/shibboleth/*.log
${LOG_DIR}/shibboleth-www/*.log
{
    rotate 4
    weekly
    missingok
    notifempty
    compress
    delaycompress
    sharedscripts
}
EOF

cat <<EOF

Run the following command to enable the logrotate configuration
    cd /etc/logrotate.d
    sudo ln -s ${LOGROTATE_DIR}/spid-auth-proxy .

EOF


example/static/js/agid-spid-enter-config.js
var config = {
    "aruba": {
        "url": "/login",
        "title": "Accedi a SPID con Aruba",
        "logo": "spid-idp-aruba.svg"
    },
    "infocert": {
        "url": "/login",
        "title": "Accedi a SPID con Infocert",
        "logo": "spid-idp-infocertid.svg"
    },
    "poste": {
        "url": "/login",
        "title": "Accedi a SPID con Poste",
        "logo": "spid-idp-posteid.svg"
    },
    "sielte": {
        "url": "/login",
        "title": "Accedi a SPID con Sielte",
        "logo": "spid-idp-sielteid.svg"
    },
    "tim": {
        "url": "/login",
        "title": "Accedi a SPID con Tim",
        "logo": "spid-idp-timid.svg"
    },
    "intesa": {
        "url": "/login",
        "title": "Fornitore non attivo",
        "logo": "spid-idp-intesaid.svg"
    },
    "register": {
        "url": "/login",
        "title": "Fornitore non attivo",
        "logo": "spid-idp-spiditalia.svg"
    },
    "namirial": {
        "url": "/login",
        "title": "Fornitore non attivo",
        "logo": "spid-idp-namirialid.svg"
    }
};


example/index.js
// Copyright 2018 AgID - Agenzia per l'Italia Digitale
//
// Licensed under the EUPL, Version 1.2 or - as soon they will be approved by
// the European Commission - subsequent versions of the EUPL (the "Licence").
//
// You may not use this work except in compliance with the Licence.
//
// You may obtain a copy of the Licence at:
//
//    https://joinup.ec.europa.eu/software/page/eupl
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the Licence is distributed on an "AS IS" basis, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// Licence for the specific language governing permissions and limitations
// under the Licence.
const express = require('express');
const path = require('path');
const session = require('express-session');

const PORT = 8080;
const SERVER_NAME = process.env.SERVER_NAME || 'localhost';
const ALLOWED_AUTHN_CONTEXTS = [
  'https://www.spid.gov.it/SpidL1',
  'https://www.spid.gov.it/SpidL2',
  'https://www.spid.gov.it/SpidL3',
];
const ATTRIBUTES = [
  'ADDRESS',
  'COMPANYNAME',
  'COUNTYOFBIRTH',
  'DATEOFBIRTH',
  'DIGITALADDRESS',
  'EMAIL',
  'EXPIRATIONDATE',
  'FAMILYNAME',
  'FISCALNUMBER',
  'GENDER',
  'IDCARD',
  'IVACODE',
  'MOBILEPHONE',
  'NAME',
  'PLACEOFBIRTH',
  'REGISTEREDOFFICE',
  'SPIDCODE',
];

const app = express();

app.use(express.static('static'));

app.use(session({
  secret: 's3cr3t',
  cookie: {
    maxAge: 3600,
  },
}));

/**
 * Index page
 */
app.get('/', (req, res) => {
  const { fiscalNumber, name } = req.session;
  if (fiscalNumber && name) {
    res.send(`You are logged as ${name} (${fiscalNumber})<br/>`
      + `<a href="https://${SERVER_NAME}/iam/Logout`
      + '?return=/logout">Logout</a>');
  } else {
    res.sendFile(path.join(__dirname + '/static/pages/smart-button.html'));
  }
});

/**
 * Login callback (/iam/Login?entityId=...&target=https://<SERVER_NAME>/login)
 */
app.get('/login', (req, res) => {
  const address = req.get('ADDRESS');
  const companyName = req.get('COMPANYNAME');
  const countyOfBirth = req.get('COUNTYOFBIRTH');
  const dateOfBirth = req.get('DATEOFBIRTH');
  const digitalAddress = req.get('DIGITALADDRESS');
  const email = req.get('EMAIL');
  const expirationDate = req.get('EXPIRATIONDATE');
  const familyName = req.get('FAMILYNAME');
  const fiscalNumber = req.get('FISCALNUMBER');
  const gender = req.get('GENDER');
  const idCard = req.get('IDCARD');
  const ivaCode = req.get('IVACODE');
  const mobilePhone = req.get('MOBILEPHONE');
  const name = req.get('NAME');
  const placeOfBirth = req.get('PLACEOFBIRTH');
  const registeredOffice = req.get('REGISTEREDOFFICE');
  const spidCode = req.get('SPIDCODE');

  const authnContext = req.get('Shib-AuthnContext-Class');

  if (!ALLOWED_AUTHN_CONTEXTS.includes(authnContext)) {
    res.send({
      error: 'Invalid AuthnContextClass',
      desc: `Value not allowed (${authnContext})`
    });
  } else {
    // NOTE: this check is aligned to the Shibboleth configuration where the
    //       comarison is set to "exact" and the default level is
    //       "https://www.spid.gov.it/SpidL1"
    if (authnContext !== 'https://www.spid.gov.it/SpidL1') {
      res.send({
        error: 'Invalid AuthnContextClass',
        desc: `Requested https://www.spid.gov.it/SpidL1 with comparison exact (${authnContext})`,
      });
    } else if (!(address || companyName || countyOfBirth || dateOfBirth
          || digitalAddress || email || expirationDate || familyName
          || fiscalNumber || gender || idCard || ivaCode || mobilePhone
          || name || placeOfBirth || registeredOffice || spidCode)) {
      res.send({
        error: 'Invalid attribute(s)',
        desc: 'No attributes were provided',
      });
    } else if (!(!address && !companyName && !countyOfBirth && !dateOfBirth
        && !digitalAddress && !email && !expirationDate && !familyName
        && fiscalNumber && !gender && !idCard && !ivaCode && !mobilePhone
        && name && !placeOfBirth && !registeredOffice && !spidCode)) {
      res.send({
        error: 'Invalid attribute(s)',
        desc: 'A different attribute set was provided',
      });
    } else if (!address && !companyName && !countyOfBirth && !dateOfBirth
        && !digitalAddress && !email && !expirationDate && !familyName
        && fiscalNumber && !gender && !idCard && !ivaCode && !mobilePhone
        && name && !placeOfBirth && !registeredOffice && !spidCode) {
      console.log(`Got login headers (${fiscalNumber}, ${name})`);
      req.session.fiscalNumber = fiscalNumber;
      req.session.name = name;
      res.redirect('/');
    } else {
      res.send({
        error: 'Error',
      });
    }
  }
});

/**
 * Logout callback (/iam/Logout?return=/logout)
 */
app.get('/logout', (req, res) => {
  req.session.destroy((err) => {
    console.log('Successfully logged out');
    res.redirect('/');
  });
});

/**
 * Free access resource
 */
app.get('/other', (req, res) => {
  res.sendFile(path.join(__dirname + '/static/pages/other.html'));
});

/**
 * Manage error query string
 * https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPErrors
 */
app.get('/error', (req, res) => {
  const { errorType, errorText, entityID, now } = req.query;
  res.send({
    entityID,
    errorText,
    errorType,
    now,
  });
});

app.listen(PORT, () => {
  console.log(`App listening on port ${PORT}`);
});
